src/plugins/discover/public/context_awareness/profiles_manager.test.ts

```diff
@@ -0,0 +1,263 @@
+      record: mocks.contextRecordMock,
+    });
+    const profiles$ = mocks.profilesManagerMock.getProfiles$({ record });
+    const next = jest.fn();
+    profiles$.subscribe(next);
+    expect(getProfilesSpy).toHaveBeenCalledTimes(1);
+    expect(next).toHaveBeenCalledWith([{}, {}, mocks.documentProfileProviderMock.profile]);
+    await mocks.profilesManagerMock.resolveRootProfile({});
+    expect(getProfilesSpy).toHaveBeenCalledTimes(2);
+    expect(next).toHaveBeenCalledWith([
+      mocks.rootProfileProviderMock.profile,
+      {},
+      mocks.documentProfileProviderMock.profile,
+    ]);
+    await mocks.profilesManagerMock.resolveDataSourceProfile({});
+    expect(getProfilesSpy).toHaveBeenCalledTimes(3);
+    expect(next).toHaveBeenCalledWith([
+      mocks.rootProfileProviderMock.profile,
+      mocks.dataSourceProfileProviderMock.profile,
+      mocks.documentProfileProviderMock.profile,
+    ]);
+  });
+
+  it("should not resolve root profile again if params haven't changed", async () => {
+    await mocks.profilesManagerMock.resolveRootProfile({ solutionNavId: 'solutionNavId' });
+    await mocks.profilesManagerMock.resolveRootProfile({ solutionNavId: 'solutionNavId' });
+    expect(mocks.rootProfileProviderMock.resolve).toHaveBeenCalledTimes(1);
+  });
+
+  it('should resolve root profile again if params have changed', async () => {
+    await mocks.profilesManagerMock.resolveRootProfile({ solutionNavId: 'solutionNavId' });
+    await mocks.profilesManagerMock.resolveRootProfile({ solutionNavId: 'newSolutionNavId' });
+    expect(mocks.rootProfileProviderMock.resolve).toHaveBeenCalledTimes(2);
+  });
+
+  it('should not resolve data source profile again if params have not changed', async () => {
+    await mocks.profilesManagerMock.resolveDataSourceProfile({
+      dataSource: createEsqlDataSource(),
+      query: { esql: 'from *' },
+    });
+    expect(mocks.dataSourceProfileProviderMock.resolve).toHaveBeenCalledTimes(1);
+    await mocks.profilesManagerMock.resolveDataSourceProfile({
+      dataSource: createEsqlDataSource(),
+      query: { esql: 'from *' },
+    });
+    expect(mocks.dataSourceProfileProviderMock.resolve).toHaveBeenCalledTimes(1);
+  });
+
+  it('should resolve data source profile again if params have changed', async () => {
+    await mocks.profilesManagerMock.resolveDataSourceProfile({
+      dataSource: createEsqlDataSource(),
+      query: { esql: 'from *' },
+    });
+    expect(mocks.dataSourceProfileProviderMock.resolve).toHaveBeenCalledTimes(1);
+    await mocks.profilesManagerMock.resolveDataSourceProfile({
+      dataSource: createEsqlDataSource(),
+      query: { esql: 'from logs-*' },
+    });
+    expect(mocks.dataSourceProfileProviderMock.resolve).toHaveBeenCalledTimes(2);
+  });
+
+  it('should log an error and fall back to the default profile if root profile resolution fails', async () => {
+    await mocks.profilesManagerMock.resolveRootProfile({ solutionNavId: 'solutionNavId' });
```

Comment:
Not saying this needs to change now, but I found these tests quite difficult to read. Contributing factors might have been:

- The error is provoked by monkey-patching a method instead of using a profile provider that fails "naturally".
- The `addLog` is an unfortunate side-effect on its own, but also using it to test formatted error messages doubles the smell.

A cleaner solution could be to decompose it into layers responsible for the cache, the async side effects and the reactivity, which could be tested separately. But I recognize that would be disproporionate effort it this point.

Reply:
Thanks for the feedback, and I'm certainly open to followup improvements (although as you mentioned, I'd prefer to limit changes at this point for the sake of unblocking dependent work). This feedback makes sense to me overall, but I'm a bit confused on this point:
>The `addLog` is an unfortunate side-effect on its own, but also using it to test formatted error messages doubles the smell.

Is your concern here the use of the static `addLog` function for logging, or that `ProfilesManager` handles this logging itself? The way I interpreted this is that you feel it would be cleaner to do something along the lines of returning or throwing an error on resolution failure, and logging it externally, rather than logging as a side effect of resolution failure, but I'm not sure if that's what you meant.

Reply:
Yes, that's exactly what I meant. But if we want to keep it in there for convenience, maybe an improvement would be to supply the logging side-effect to the profile manager as a constructor argument? :thinking: